src/main/java/duke/ui/Ui.java

```diff
@@ -0,0 +1,202 @@
+
+    @FXML
+    private ScrollPane scrollPane;
+    @FXML
+    private VBox dialogContainer;
+    @FXML
+    private TextField userInput;
+    @FXML
+    private Button sendButton;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/DaDuke.png"));
+
+    private DukeChatbot dukeChatbot;
+    private MessageFormatter messageFormatter;
+
+    /**
+     * Creates a Ui instance.
+     */
+    public Ui() {
+        this.messageFormatter = MessageFormatter.getInstance();
+    }
+
+    /**
+     * Displays the greeting.
+     */
+    public void printGreeting() {
+        printFormattedMessage(GREETING_MESSAGE);
+    }
+
+    /**
+     * Displays the generic error message for unexpected errors.
+     */
+    public void printUnexpectedErrorMessage() {
+        printFormattedMessage(UNEXPECTED_ERROR_MESSAGE);
+    }
+
+    /**
+     * Displays the error message for when an invalid command occurs.
+     *
+     * @param message The error message.
+     */
+    public void printInvalidCommandErrorMessage(String message) {
+        printFormattedMessage(String.format(INVALID_COMMAND_ERROR_TEMPLATE, message));
+    }
+
+    /**
+     * Displays the exit message.
+     */
+    public void printExitMessage() {
+        printFormattedMessage(EXIT_MESSAGE);
+    }
+
+    /**
+     * Initiates the message-building process.
+     *
+     * @return The MessageBuilder instance used in this message-building process.
+     */
+    public MessageBuilder startMessage() {
+        return new MessageBuilder();
+    }
+
+    public void setDukeChatbot(DukeChatbot dukeChatbot) {
+        this.dukeChatbot = dukeChatbot;
+    }
+
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    /**
+     * Processes the user's input.
+     * Creates one dialog box echoing user input and then appends them to the dialog container.
+     * Adapted from https://se-education.org/guides/tutorials/javaFxPart4.html written by Jeffry Lum.
+     */
+    @FXML
+    private void handleUserInput() {
+        String input = userInput.getText();
+        dialogContainer.getChildren().add(DialogBox.getUserDialog(input, userImage));
+        userInput.clear();
+        dukeChatbot.readInput(input);
+    }
+
+    private String getListLengthMessage(int listLength) {
+        // Check whether singular or plural should be printed.
+        if (listLength != 1) {
+            return String.format("Now you have %d tasks in the list.", listLength);
+        } else {
+            return "Now you have 1 task in the list.";
+        }
+    }
+
+    private void printFormattedMessage(String message) {
+        String formattedMessage = messageFormatter.getFormattedMessage(message);
+        dialogContainer.getChildren().add(DialogBox.getDukeDialog(formattedMessage, dukeImage));
```

Comment:
Would this be better if it is broken down to different lines? (to improve readability)

Reply:
Yes, I agree. I must have overlooked that. Thank you for pointing it out!